lib/transaction_files.c

```diff
@@ -0,0 +1,446 @@
+		oldpkgname = xbps_pkg_name(item->old.pkgver);
+		if (strcmp(newpkgname, oldpkgname) != 0) {
+			if (remove) {
+				xbps_dbg_printf(xhp, "%s: [trans] file `%s' moved to"
+				    " package `%s'\n", pkgver, file, item->new.pkgver);
+			} else {
+				xbps_dbg_printf(xhp, "%s: [trans] file `%s' moved from"
+				    " package `%s'\n", pkgver, file, item->new.pkgver);
+			}
+		}
+		free(newpkgname);
+		free(oldpkgname);
+	}
+
+	return 0;
+}
+
+static int
+collect_files(struct xbps_handle *xhp, xbps_dictionary_t d,
+			const char *pkgver, bool remove)
+{
+	struct stat st;
+	xbps_array_t a;
+	xbps_dictionary_t filed;
+	uint64_t size;
+	unsigned int i;
+	int rv = 0;
+	const char *file;
+
+	if ((a = xbps_dictionary_get(d, "files"))) {
+		for (i = 0; i < xbps_array_count(a); i++) {
+			filed = xbps_array_get(a, i);
+			xbps_dictionary_get_cstring_nocopy(filed, "file", &file);
+			size = 0;
+			xbps_dictionary_get_uint64(filed, "size", &size);
+			rv = collect_file(xhp, file, size, pkgver, TYPE_FILE, remove);
+			if (rv != 0)
+				goto out;
+		}
+	}
+	if ((a = xbps_dictionary_get(d, "conf_files"))) {
+		for (i = 0; i < xbps_array_count(a); i++) {
+			filed = xbps_array_get(a, i);
+			xbps_dictionary_get_cstring_nocopy(filed, "file", &file);
+			size = 0;
+			xbps_dictionary_get_uint64(filed, "size", &size);
+			/* XXX: how to handle conf_file size */
+			if (remove && stat(file, &st) != -1 && size != (uint64_t)st.st_size)
+				size = 0;
+			rv = collect_file(xhp, file, size, pkgver, TYPE_FILE, remove);
+			if (rv != 0)
+				goto out;
+		}
+	}
+	if ((a = xbps_dictionary_get(d, "links"))) {
+		for (i = 0; i < xbps_array_count(a); i++) {
+			filed = xbps_array_get(a, i);
+			xbps_dictionary_get_cstring_nocopy(filed, "file", &file);
+			rv = collect_file(xhp, file, 0, pkgver,  TYPE_LINK, remove);
+			if (rv != 0)
+				goto out;
+		}
+	}
+	if ((a = xbps_dictionary_get(d, "dirs"))) {
+		for (i = 0; i < xbps_array_count(a); i++) {
+			filed = xbps_array_get(a, i);
+			xbps_dictionary_get_cstring_nocopy(filed, "file", &file);
+			rv = collect_file(xhp, file, 0, pkgver,  TYPE_DIR, remove);
+			if (rv != 0)
+				goto out;
+		}
+	}
+
+out:
+	return rv;
+}
+
+static int
+add_from_archive(struct xbps_handle *xhp, xbps_dictionary_t pkg_repod)
+{
+	xbps_dictionary_t filesd;
+	struct archive *ar = NULL;
+	struct archive_entry *entry;
+	struct stat st;
+	const char *pkgver;
+	char *bpkg;
+	/* size_t entry_size; */
+	int rv = 0, pkg_fd = -1;
+
+	xbps_dictionary_get_cstring_nocopy(pkg_repod, "pkgver", &pkgver);
+	assert(pkgver);
+
+	bpkg = xbps_repository_pkg_path(xhp, pkg_repod);
+	if (bpkg == NULL)
+		return errno;
+
+	if ((ar = archive_read_new()) == NULL) {
+		free(bpkg);
+		return ENOMEM;
+	}
+
+	/*
+	 * Enable support for tar format and gzip/bzip2/lzma compression methods.
+	 */
+	archive_read_support_compression_gzip(ar);
+	archive_read_support_compression_bzip2(ar);
+	archive_read_support_compression_xz(ar);
+	archive_read_support_format_tar(ar);
```

Comment:
I think the number of times we do this is embarrassing. We'd need a private function to remove redundant code.

Also please change them to `archive_read_filter_compression_xx`, the `archive_read_support_xxx` are from the old interface.